pgmpy/estimators/NOTEARS.py

```diff
@@ -0,0 +1,420 @@
+        # The penalty matrix is: $ P_{ij} e^{-(w_t - W^t_{ij} * M_{ij})^{- \alpha}} $
+        # The penalty is $ \sum_{ij} P_{ij} $.
+        # The Jacobian is $ J_{ij} = -\alpha M_{ij} P_{ij} (w_t -  W_{ij} M_{ij})^{- \alpha - 1} $.
+        # changes?
+
+        W_thres = self.backend.where(
+            adjacency_matrix < weight_threshold, adjacency_matrix, 0
+        )
+        penalty_mat = self.backend.exp(
+            -((weight_threshold - (W_thres * fixed_mask)) ** (-alpha))
+        )
+        jac = (
+            -alpha
+            * fixed_mask
+            * penalty_mat
+            * (weight_threshold - W_thres * fixed_mask) ** (-alpha - 1)
+        )
+
+        return self.backend.sum(penalty_mat), jac
+
+    def _objective_function(
+        self,
+        adjacency_matrix_doubled,
+        alpha,
+        rho,
+        lambda1,
+        lambda2,
+        lambda3,
+        loss_type,
+        data,
+        forbidden_mask,
+        fixed_mask,
+        w_threshold,
+        alpha_2,
+    ):
+        """
+        Evaluate value and gradient of augmented Lagrangian for doubled variables ([2 d^2] array).
+
+        Parameters
+        ----------
+
+        Returns
+        -------
+        obj: float
+            The value of the objective function for given parameters.
+
+        g_obj: numpy.ndarray
+            The gradient of the objective function for given parameters.
+        """
+        adjacency_matrix_est = self._adj(adjacency_matrix_doubled)
+        loss, loss_jac = self._loss_grad(loss_type, data, adjacency_matrix_est)
+        acyclic_penalty, acyclic_jac = self._constraint_grad(adjacency_matrix_est)
+
+        forb_penalty, forb_jac = self._forbidden_penalty_gradient(
+            adjacency_matrix_est, forbidden_mask
+        )
+        fixed_penalty, fixed_jac = self._fixed_penalty_gradient(
+            adjacency_matrix_est, fixed_mask, alpha_2, w_threshold
+        )
+        # print(forb_penalty, forb_jac)
+
+        obj = (
+            loss
+            + 0.5 * rho * acyclic_penalty * acyclic_penalty
+            + alpha * acyclic_penalty
+            + lambda1 * adjacency_matrix_doubled.sum()
+            + lambda2 * forb_penalty
+            + lambda3 * fixed_penalty
+        )
+        G_smooth = (
+            loss_jac
+            + (rho * acyclic_penalty + alpha) * acyclic_jac
+            + lambda2 * forb_jac
+            + lambda3 * fixed_jac
+        )
+
+        g_obj = self.backend.concatenate(
+            (
+                G_smooth + lambda1 + lambda2 * forb_jac + lambda3 * fixed_jac,
+                -G_smooth + lambda1 + lambda2 * forb_jac + lambda3 * fixed_jac,
+            ),
+            axis=None,
+        )
+        return obj, g_obj
+
+    def estimate(
+        self,
+        lambda1,
+        lambda2=10,
+        lambda3=10,
+        loss_type="l2",
+        max_iter=20,
+        h_tol=1e-8,
+        rho_max=1e16,
+        w_threshold=0.3,
+        c=0.25,
+        alpha_2=1,

```

Comment:
We will need to reduce the number of hyperparameters. This is not user friendly and would be very difficult for anyone to use it.